src/lib/logical_query_plan/abstract_lqp_node.cpp

```diff
@@ -274,60 +274,80 @@ bool AbstractLQPNode::has_matching_unique_constraint(const ExpressionUnorderedSe
 }
 
 std::vector<FunctionalDependency> AbstractLQPNode::functional_dependencies() const {
-  // Gather input FDs from all input nodes
-  auto fds_left = std::vector<FunctionalDependency>();
-  auto fds_right = std::vector<FunctionalDependency>();
-  if (left_input()) {
-    fds_left = left_input()->functional_dependencies();
-  }
-  if (right_input()) {
-    fds_right = right_input()->functional_dependencies();
+  // (1) Gather FDs from previous nodes
+  auto fds_in = pass_functional_dependencies();
+  if constexpr (HYRISE_DEBUG) {
+    auto distinct_fds = std::vector<FunctionalDependency>();
+    for (const auto& fd : fds_in) {
+      Assert(std::none_of(distinct_fds.cbegin(), distinct_fds.cend(),
+                          [&fd](const auto& distinct_fd) { return fd == distinct_fd; }),
+             "Did not expect LQP nodes to pass duplicate FDs.");
+      distinct_fds.push_back(fd);
```

Comment:
`DebugAssert(std::set(fds_in.begin(), fds_in.end()).size() == fds_in.size())`?

Reply:
Thanks, very clever!
When having an `if constexpr (HYRISE_DEBUG) { }`-block, I always use regular Asserts. Shall I use DebugAssert anyway?

Reply:
No hard feelings about it. Assert might be better because you might run into unused variables if you comment out the `if constexpr`. Is it still necessary here though?

Reply:
No, it is just my personal practice.

Reply:
Which is good. My point was that the entire block should become redundant with the std::set hack from above

Reply:
```cpp
  if constexpr (HYRISE_DEBUG) {
    const auto distinct_fds = std::unordered_set<FunctionalDependency>(fds_in.cbegin(), fds_in.cend());
    Assert(distinct_fds.size() == fds_in.size(), "Did not expect LQP nodes to pass duplicate FDs.");
  }
```

I stick with the block to remove the variable from release builds.

Reply:
ok ¯\\\_(ツ)\_/¯